### ISSUE TITLE
Fix node input concurrency deadlock

### ISSUE THREAD

**Issue by Pwuts** [Significant-Gravitas/AutoGPT#7935]
Originally reported [on Discord](https://discord.com/channels/1092243196446249134/1278724007805452411/1279560842446110771):

> I'm seeing another issue now where `register_next_executions(..)` fails to get a lock on the node input. I've established that the lock request isn't dropped by Pyro, but I have no idea why *sometimes* it just isn't given to a waiting lock request.
>
> It's like:
>
> 1. Worker A, B and C request lock
> 2. 🔒 Worker A gets lock
> 3. 🔓 Worker A releases lock
> 4. 🔒 Worker B gets lock
> 5. 🔓 Worker B releases lock
> 6. Worker D requests lock
> 7. 🔒 Worker D gets lock
>
> ... worker C is still waiting, and will wait forever.
>
> The order of steps 5 and 6 can be the other way around too btw, result remains the same.
>
> The weird thing is that this only occurs during `test_static_input_link_on_graph`.

This issue is causing `test_static_input_link_on_graph` to be flaky, and will cause deadlocks in any graph execution where >2 nodes try to send input to the same node around the same time.

**Comment by Pwuts** [comment]
This turns out to be an issue with the `KeyedMutex`, which deletes its lock entries on release, causing the next release *not to happen*:

```py
    def unlock(self, key: Any):
        with self.locks_lock:
            if key in self.locks:
                lock = self.locks.pop(key)
            else:
                return
        lock.release()
```

So:

1. Worker A, B and C request lock (creating new lock object 1)
2. 🔒 Worker A gets lock (lock 1)
3. 🔓 Worker A releases lock (lock 1, which is deleted from `locks`)
4. 🔒 Worker B gets lock (lock 1, which it was already waiting on)
5. Worker B attempts to release lock, but since the key is no longer in `locks`, it returns before releasing the lock, leaving the lock unreleased
6. Worker C waits forever
